File: search/website_search.py
"""
website_search.py — Company website source adapter.

Crawls company websites that appear in search results for
Singing Bowls importers and extracts contact emails from
their contact/about pages.
"""

import logging
import re
import time
from typing import List, Dict
from urllib.parse import urljoin, urlparse

import requests
from bs4 import BeautifulSoup

from config import SEARCH_KEYWORD, MAX_RESULTS_PER_SOURCE, BLOCKLISTED_DOMAINS, IMAGE_EXTENSIONS

logger = logging.getLogger(__name__)

HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/120.0.0.0 Safari/537.36"
    ),
    "Accept-Language": "en-US,en;q=0.9",
}

EMAIL_PATTERN = re.compile(r"[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}")

# Seed URLs — well-known B2B platforms for Singing Bowls buyers
SEED_URLS = [
    "https://www.alibaba.com/trade/search?fsb=y&IndexArea=product_en&CatId=&SearchText={keyword}+buyer",
    "https://www.made-in-china.com/products-search/hot-china-products/{keyword}.html",
    "https://www.hktdc.com/sourcing/hong-kong-international-trade-fair.htm",
    "https://www.kompass.com/a/singing-bowls/",
    "https://directory.fairtradeusa.org/",
]

# Contact page path patterns to try on each discovered domain
CONTACT_PATHS = ["/contact", "/contact-us", "/about", "/about-us", "/info", "/reach-us"]


def _is_blocked(url: str) -> bool:
    for domain in BLOCKLISTED_DOMAINS:
        if domain in url:
            return True
    return False


def _extract_emails(text: str) -> List[str]:
    return list({
        e.lower() for e in EMAIL_PATTERN.findall(text)
        if not any(e.lower().endswith(ext) for ext in IMAGE_EXTENSIONS)
        and len(e.split("@")[-1]) <= 50
    })


def _extract_mailto(soup: BeautifulSoup) -> List[str]:
    emails = []
    for link in soup.select("a[href^='mailto:']"):
        href = link.get("href", "")
        email = href.replace("mailto:", "").split("?")[0].strip().lower()
        if email and "@" in email:
            emails.append(email)
    return emails


def _crawl_contact_pages(base_url: str) -> List[str]:
    """Try known contact page paths on a domain and extract emails."""
    emails = []
    parsed = urlparse(base_url)
    base = f"{parsed.scheme}://{parsed.netloc}"

    for path in CONTACT_PATHS:
        try:
            url = base + path
            resp = requests.get(url, headers=HEADERS, timeout=8)
            if resp.status_code == 200:
                soup = BeautifulSoup(resp.text, "lxml")
                emails += _extract_mailto(soup)
                emails += _extract_emails(soup.get_text())
                if emails:
                    break
            time.sleep(0.5)
        except Exception:
            pass

    return list(set(emails))


def _scrape_seed(url: str, keyword: str) -> List[Dict]:
    """Scrape a seed URL and extract company/buyer records."""
    records = []
    try:
        full_url = url.format(keyword=requests.utils.quote(keyword))
        resp = requests.get(full_url, headers=HEADERS, timeout=12)
        soup = BeautifulSoup(resp.text, "lxml")

        # Collect all outbound links to company websites
        seen_domains = set()
        for link in soup.select("a[href]"):
            href = link.get("href", "")
            if not href.startswith("http"):
                href = urljoin(full_url, href)
            parsed = urlparse(href)
            domain = parsed.netloc
            if not domain or _is_blocked(href) or domain in seen_domains:
                continue
            seen_domains.add(domain)

            # Crawl the company's contact pages
            emails = _crawl_contact_pages(href)
            company_text = link.get_text(strip=True)[:80]

            for email in emails[:2]:
                records.append({
                    "buyer_name": "",
                    "company_name": company_text,
                    "email": email,
                    "website": href,
                    "country": "",
                    "source_platform": "Website",
                })

            if len(records) >= 5:
                break

        time.sleep(1)
    except Exception as exc:
        logger.warning("Website seed scrape failed for %s: %s", url, exc)

    return records


def search(keyword: str = SEARCH_KEYWORD, max_results: int = MAX_RESULTS_PER_SOURCE) -> List[Dict]:
    """
    Main entry point for the company website source adapter.
    Returns buyer record dicts.
    """
    logger.info("Website adapter: starting search for '%s'", keyword)
    all_records: List[Dict] = []

    for seed_url in SEED_URLS:
        if len(all_records) >= max_results:
            break
        records = _scrape_seed(seed_url, keyword)
        all_records.extend(records)

    logger.info("Website adapter: found %d records", len(all_records))
    return all_records[:max_results]

File: src/lib/search/websiteSearch.ts
import { SEARCH_KEYWORD, MAX_RESULTS_PER_SOURCE, BLOCKLISTED_DOMAINS } from '@/lib/config';
import type { BuyerRecord } from '@/types';
import { fetchWithScrapingBee, searchGoogleApi } from '@/lib/search/scrapingBee';

const EMAIL_PATTERN = /[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/g;

const WEBSITE_QUERIES = [
  '"{keyword}" importer wholesale "contact us" email',
  '"{keyword}" import company "info@" OR "sales@"',
];

function isBlocked(url: string): boolean {
  for (const domain of BLOCKLISTED_DOMAINS) {
    if (url.includes(domain)) return true;
  }
  return false;
}

async function visitContactPage(baseUrl: string): Promise<string[]> {
  const contactPaths = ['/contact', '/contact-us', '/'];
  for (const p of contactPaths) {
    try {
      const url = new URL(p, baseUrl).toString();
      const resp = await fetchWithScrapingBee(url, false, 8000);
      const html = await resp.text();
      const emails = html.match(EMAIL_PATTERN) ?? [];
      if (emails.length) return emails.map((e) => e.toLowerCase());
    } catch {
      // try next path
    }
  }
  return [];
}

export async function search(
  keyword = SEARCH_KEYWORD,
  maxResults = MAX_RESULTS_PER_SOURCE
): Promise<Partial<BuyerRecord>[]> {
  const results: Partial<BuyerRecord>[] = [];
  const seenUrls = new Set<string>();

  for (const template of WEBSITE_QUERIES) {
    const query = template.replace('{keyword}', keyword);
    const searchResults = await searchGoogleApi(query, maxResults);
    
    const pagePromises: Promise<void>[] = [];

    for (const res of searchResults) {
      const title = res.title;
      const href = res.url;
      const snippet = res.snippet;

      if (!href || !href.startsWith('http') || isBlocked(href) || seenUrls.has(href)) continue;
      seenUrls.add(href);

      // First check snippet for emails
      const snippetEmails = (snippet.match(EMAIL_PATTERN) ?? []).map((e) => e.toLowerCase());
      if (snippetEmails.length) {
        results.push({
          buyer_name: title,
          company_name: title,
          email: snippetEmails[0],
          website: href,
          country: '',
          source_platform: 'Website',
          discovered_at: new Date().toISOString(),
        });
        continue;
      }

      // Otherwise, visit the site
      const p = visitContactPage(href).then((emails) => {
        if (emails.length) {
          results.push({
            buyer_name: title,
            company_name: title,
            email: emails[0],
            website: href,
            country: '',
            source_platform: 'Website',
            discovered_at: new Date().toISOString(),
          });
        }
      });
      pagePromises.push(p);
    }

    await Promise.allSettled(pagePromises);
    if (results.length >= maxResults) break;
  }

  return results.slice(0, maxResults);
}
